### ISSUE TITLE
Freeze Transformer Weight

### ISSUE THREAD

**Issue by ivokun** [kingoflolz/mesh-transformer-jax#157]
Can we finetune the model with Transformer layers' weight freeze (only train the embeddings and softmax)?

**Comment by vfbd** [comment]
Yeah, modify the `grad` variable after here
https://github.com/kingoflolz/mesh-transformer-jax/blob/0f9b555fc1e7f8a94e77347b46dedaae5045811d/mesh_transformer/transformer_shard.py#L147
and after here
https://github.com/kingoflolz/mesh-transformer-jax/blob/0f9b555fc1e7f8a94e77347b46dedaae5045811d/mesh_transformer/transformer_shard.py#L155
so that it only contains the haiku module parameters you want; e.g. if you only want `causal_transformer_shard/~/embedding_shard/~/linear` (the embedding parameters), you can put
```python
grad = grad["causal_transformer_shard/~/embedding_shard/~/linear"]
```
after both of the lines.

If you do that, you'll also have to re-initialize the optimizer by putting this after the `network = CausalTransformer(params)` line in your training script:
```python
def _init_opt_state(params, aux):
    return network.config["optimizer"].init(params)
init_opt_state = jax.experimental.maps.xmap(
    fun=_init_opt_state,
    in_axes=(["shard", ...], ["batch", ...]),
    out_axes=["shard", ...],
    donate_argnums=(0,),
    axis_resources={"shard": "mp", "batch": "dp"}
)
network.state["opt_state"] = init_opt_state(
    network.state["params"]["causal_transformer_shard/~/embedding_shard/~/linear"],
    np.empty(1)
)
```

**Comment by ivokun** [comment]
Great! I'll give it a try!
Thank you so much for your thorough answer!

**Comment by paramjeet2021** [comment]
Has anyone tried these updates and let me know if it worked?
